docs/guides/optimizers.rst

```diff
@@ -4,7 +4,398 @@
+Optimization (i.e. ability to minimize a scalar function) lies at the heart of scientific computing
+and its applications. For instance, being able to minimize a cost function is crucial for
+Variational Quantum Algorithms (VQAs).
 
-- The typical arguments to an optimizer. They are usually simple and don't require other objects.
-- 
\ No newline at end of file
+Optimization is nothing new, and there exists plethora of ready to use libraries with various
+optimizers. So what new can Orquestra bring to the table? The answer, as usually with Orquestra,
+is unification. The `orquestra-opt` package defines interfaces for all things related to
+optimization and provides implementation of many existing optimizers.
+
+
+What this guide covers
+======================
+This guide will introduce you to everything you need to know to optimize scalar functions using
+`orquestra-opt` package. In particular we'll discuss:
+
+- What functions can be minimized? Definition of cost function.
+- Gradients, and how to add them to your cost function.
+- Orquestra's approach for implementing optimization algorithms - the
+  :class:`Optimizer <orquestra.opt.api.Optimizer>` interface.
+- How to record history of evaluation of the cost function.
+- List of optimizers are currently available.
+
+
+Cost functions
+==============
+
+Before we start optimizing, we need to know what can be optimized. In Orquestra, we deal only with
+minimization of functions of the form :math:`f:\mathbb{R}^N \to \mathbb{R}`. Throughout the
+codebase, such functions are called the *cost functions*.
+
+All optimizers expect that the arguments to cost functions are stored as a single vector in a 1-D
+numpy arrays. Suppose you want to create the following cost function:
+:math:`f(x, y, z) = x^2 + y^2 + z^2`. Out of the two natural ways of defining it in Python, the
+second one is the one expected by Optimizers:
+
+.. literalinclude:: ../examples/guides/optimizers.py
+  :language: python
+  :start-after: # Definition of cost function
+  :end-before: # --- End
+
+.. note::
+   Remember that the argument to your cost function should be called `parameters`. While
+   it makes no difference for most optimizers at runtime, it makes your function match
+   structurally to the :class:`CostFunction <orquestra.opt.api.CostFunction>` protocol expected
+   by Optimizers.
+
+Adding gradients
+================
+Some optimization methods use gradients to move across the optimization landscape in the
+correct direction. You can enrich your cost function with gradient in several ways.
+
+To start with, cost function with gradients are just normal cost functions having additional
+`gradient` attribute. This attribute should also be a callable returning gradient
+of you function at a given point. Hence, one possible way of defining a cost function with
+gradient is to just write a class with `__call__` dunder method and a gradient attribute.
```

Comment:
```suggestion
gradient is to just write a class with `__call__` method and a gradient attribute.
```

I'd drop `dunder` from the text, might be confusing if someone doesn't know what it is (and I guess a lot of people won't)